### ISSUE TITLE
[RFE] Scrape metrics from multiple prom sources

### ISSUE THREAD

**Issue by jtaleric** [cloud-bulldozer/kube-burner#245]
# Use case
With hypershift, there are at least two sources of metrics we are interested in (possibly 3). 

- Host Cluster (HC) metrics, which will be the Worker nodes.
- Hosted Cluster Services, which will be running on the Management Cluster  (user workload metrics - uwm)
- (Optional) Management Cluster Prometheus

# kube-burner to collect metrics from multiple prom sources
Today, kube-burner (in an OCP world) will discover the Prometheus resources, and generate an auth token for scraping. This makes collecting metrics simple and efficient. However, in an hypershift world, there are many sources to collect from. 

# Alternatives
Using `remote_write` have prometheus send metrics to Thanos, and have Kube-burner collect metrics from Thanos instead of Prometheus.

This is effective, but requires Thanos infrastructure. To reduce the barrier, we should look at the feasibility to have kube-burner collect these metrics independently.

**Comment by rsevilla87** [comment]
Thanks for bringing this up @jtaleric , as we've discussed several times, we should be able to get away from thanos, because of several reasons I'm not going to enumerate here :)

As an alternative, I think we can leverage the kube-burner's `index` subcommand to collect HCP metrics exposed by the UWM stack. This subcommand receives a start/end timestamp to scratch metrics an specific time period. Hence a valid approach could be:

```shell
# Run workload against HC and let kube-burner collect/index metrics from it
$ start=$(date +%s)
$ kube-burner init -c <workload> --uuid= ${UUID} -m metrics.yml -u=hc-prometheus.endpoint.com
$ end=$(date +%s)
# Run the mentioned kube-burner index tool to collect metrics from the UWM stack of the management cluster
$ kube-burner index -c <workload> --uuid= ${UUID} -m metrics.yml -u=uwm-prometheus.endpoint.com --start=${start} --end=${end}
```

By using this mechanism we should consolidate metrics from the HCP and HC.

cc: @dry923 @mukrishn @smalleni

**Comment by jtaleric** [comment]
> Thanks for bringing this up @jtaleric , as we've discussed several times, we should be able to get away from thanos, because of several reasons I'm not going to enumerate here :)

Yeah - I wasn't suggesting we keep Thanos, I was just calling out the alternative. 

> As an alternative, I think we can leverage the kube-burner's index subcommand to collect HCP metrics exposed by the UWM stack. This subcommand receives a start/end timestamp to scratch metrics an specific time period. Hence a valid approach could be

This is another possible alternative, but we would need to test it out.

**Comment by jtaleric** [comment]
cc @graflus

**Comment by jtaleric** [comment]
So, I have been kicking the tires w/ what @rsevilla87 mentioned above. (using `init` with `index`.

```
export start_ms=$(date +%s%3N) 
export start=$(date +%s) 
export UUID=$(uuidgen)
JOB_ITERATIONS=500 /root/rook/kube-burner/bin/amd64/kube-burner init \
-c cluster-density.yml \
-u https://prometheus-k8s-openshift-monitoring.apps.bm.example.com \  # Management Cluster Prom
-t $M_TOKEN \                                                         # Token for the management cluster 
--skip-tls-verify \
-m mc-metrics.yaml  \ # Check out my example here https://gist.github.com/jtaleric/23b781339b82f1ba6dba2b04c87db2f3
--log-level info \
--uuid $UUID 

export end=$(date +%s) # Workload complete 

JOB_ITERATIONS=0 ~/rook/kube-burner/bin/amd64/kube-burner index \
-u https://localhost:9091 \
-t $TOKEN \
--uuid $UUID \
--skip-tls-verify \
-m metrics-hc.yaml \   # Check out my example here : https://gist.github.com/jtaleric/23b781339b82f1ba6dba2b04c87db2f3 
-c cluster-density.yml \
--start $start \
--end $end

/metadata.sh cluster_density $start_ms $(date +%s%3N) # This is a simple script which does similar to this function https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/utils/common.sh#L120 
```

The first execution is the actual workload. The second execution just prom collection during the test duration, from the Hosted Cluster. The last script is the metadata that our Dashboards depend on. 

From the screenshot below, we can see most of the metrics are present. However since there is no "master" nodes in hypershift, those panels are empty. 

![image](https://user-images.githubusercontent.com/3965163/215192291-66fa1329-68c6-4fb0-a7a0-5ce1eea849cb.png)

One thing we should consider is different Dashboard profiles.

- Performance Engineer who is focused on OCP as a Platform (Possibly reuse the current Kube-burner Report dashboard w/o the "Master" Panels) 
- Performance Engineer who is focused on Hypershfit as a platform -- These would be more tailored to collect Management Cluster + some HC Data

I am sure those are other profiles, just kicking around the idea.

**Comment by mukrishn** [comment]
Linking [Slackthread](https://redhat-internal.slack.com/archives/GTHU29259/p1674569402179079?thread_ts=1674564942.460849&cid=GTHU29259)

TLDR: Yes it works but need some work, we have been using `kube-burner index` already on our hypershift airflow scripts for repotring metrics during [installation](https://github.com/cloud-bulldozer/airflow-kubernetes/blob/master/dags/openshift_nightlies/scripts/install/hypershift.sh#L311) and it is using Thanos database using a single master [metric profile](https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner/metrics-profiles/hypershift-metrics.yaml). 

Also we have an existing hypershift specific dashboard - [Hypershift report Imported](https://grafana.rdu2.scalelab.redhat.com:3000/d/hcp/hypershift-report-imported) that focuses hypershift performance monitoring for both management cluster as well as Hostedcluster metrics, we can re-use this dashboard.

As of now they have been dependent on Thanos but it can be modified to work the other way around as @rsevilla87 mentioned but it needs 3 metric different profiles pointing 3 prometheus endpoints and indexing to ES on a same UUID per execution.
1. Management Prometheus data - Cluster level metric 
2. Management UWM data - HCP metrics (without platform monitoring) 
3. HostedCluster prometheus data - Cluster level metric

**Comment by jtaleric** [comment]
> Linking [Slackthread](https://redhat-internal.slack.com/archives/GTHU29259/p1674569402179079?thread_ts=1674564942.460849&cid=GTHU29259)
> 
> TLDR: Yes it works but need some work, we have been using `kube-burner index` already on our hypershift airflow scripts for repotring metrics during [installation](https://github.com/cloud-bulldozer/airflow-kubernetes/blob/master/dags/openshift_nightlies/scripts/install/hypershift.sh#L311) and it is using Thanos database using a single master [metric profile](https://github.com/cloud-bulldozer/e2e-benchmarking/blob/master/workloads/kube-burner/metrics-profiles/hypershift-metrics.yaml).
> 
> Also we have an existing hypershift specific dashboard - [Hypershift report Imported](https://grafana.rdu2.scalelab.redhat.com:3000/d/hcp/hypershift-report-imported) that focuses hypershift performance monitoring for both management cluster as well as Hostedcluster metrics, we can re-use this dashboard.
> 
> As of now they have been dependent on Thanos but it can be modified to work the other way around as @rsevilla87 mentioned but it needs 3 metric different profiles pointing 3 prometheus endpoints and indexing to ES on a same UUID per execution.
> 
> 1. Management Prometheus data - Cluster level metric
> 2. Management UWM data - HCP metrics (without platform monitoring)
> 3. HostedCluster prometheus data - Cluster level metric

Agreed. 

```
Management Prometheus data - Cluster level metric
Management UWM data - HCP metrics (without platform monitoring)
HostedCluster prometheus data - Cluster level metric
```

IMHO this is pretty trivial to setup. The dashboards that currently are built on Thanos *might* need some tweaking, or possibly we could reuse the dashboards by using the same metric lable/name

**Comment by jtaleric** [comment]
Based on some of the feedback here, and my attempts to see if this is feasable, it would be good to define a new `metric-endpoint.yaml` which the user can define 1 or many Prometheus URLs. 

```yaml
- endpoint: url1
  token: token1
  profile: profile.yaml
- endpoint: url2
  token: token2
  profile: profile2.yaml
... 
```

The user would pass this to `kube-burner` to scrape from the different metric endpoints.  Thoughts?

**Comment by rsevilla87** [comment]
> 

It could make some sense, however, does it worth the effort to include a feature like this in kube-burner? I'm saying so because this  corner case "problem" can be easily fixed with some bash logic.

**Comment by jtaleric** [comment]
> > 
> 
> 
> 
> It could make some sense, however, does it worth the effort to include a feature like this in kube-burner? I'm saying so because this  corner case "problem" can be easily fixed with some bash logic. IMO kube-burner should add some extra value to justify this addition. What do you think?
👍 
IMHO maintaining this outside of kube-burner in a set of bash scripts or some other scripting language lends itself to corner cases too. 

If we all agree maintaining this outside of kube-burner is the preferred path, I am good with that. 

However, I look at our latest transitions: 

- kube-burner via benchmark-operator 
- kube-burner via e2e
- kube-burner without higher level scripting or orchestration. (ocp wrapper) 

thoughts?

**Comment by jtaleric** [comment]
> If the target application of this RFE is hypershift, does it make sense to build this into a "hypershift wrapper" in kube-burner?

Initially, yes.

> 
> Sounds to me like if the discovery required for hypershift is entirely different than that for OCP, then it can be another implementation to do the needful.

+1
